### PR TITLE
  Add DMU (Domain Mixed Unit) layer for arithmetic operations

### PR DESCRIPTION
**Pull request by marict**
This PR adds the Domain Mixed Unit (DMU) layer to enable replication of the experimental results in the Domain Mixed Unit paper submitted to arXiv:

  https://arxiv.org/abs/2509.08180

  ## Changes
  - **Add `stable_nalu/layer/dmu.py`**: Complete DMU layer 
  implementation with domain mixing logic
  - **Update `stable_nalu/layer/__init__.py`**: Export DMULayer 
  for import
  - **Update `stable_nalu/layer/generalized.py`**: Register DMU 
  in layer mapping

  ## Features
  The DMU layer implements a learnable domain-mixed arithmetic 
  unit that:
  - Blends linear and logarithmic computation domains
  - Uses learnable gating parameters to mix between domains
  - Supports basic arithmetic operations (add, sub, mul, div)

  ## Usage
  ```bash
  python experiments/single_layer_benchmark.py \
      --layer-type DMU \
      --operation <add|mul|sub|div> \
      --input-size 2 \
      --batch-size 512 \
      --max-iterations 3000 \
      --learning-rate 1e-2 \
      --no-cuda

### CODE REVIEW

**Comment by marict** [comment]
@bmistry4 

Hello! Would you be interested in reviewing this new work? Happy to share what I've been working on. Thanks! :)

**Comment by bmistry4** [comment]
Hi @marict! 

Thanks for using and adding to the benchmark! 

Happy to review the PR, though please note, it may take a bit of time since the repo is not a primary focus and I’ll need to review it in my spare time.

File: stable_nalu/layer/dmu.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        # Handle parameter freezing for freeze_G mode
+        if self.freeze_G:
+            # Create a custom parameter that only allows gradients for the second element
+            self._setup_freeze_G_gradients()
+
+        self._mag_min = 1e-11
+        self._mag_max = 1e6
+        self._log_lim = math.log(self._mag_max) - 1.0
+
+    def _setup_freeze_G_gradients(self):
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:96]: Remove function as it's redundant

File: stable_nalu/layer/dmu.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+import math
+import pdb
+
+import torch
+import torch.nn as nn
+
+from ..abstract import ExtendedTorchModule
+
+class DMULayer(ExtendedTorchModule):
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:11]: Since the layer is implemented to only support 2 inputs and 1 output, modify the class name to reflect this e.g., DMU2In1OutLayer.

File: stable_nalu/layer/dmu.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        self.op = op
+        self.freeze_G = bool(freeze_G)
+        self.freeze_O = bool(freeze_O)
+        self.unfreeze_eval = bool(unfreeze_eval)
+        self.G_perturbation = float(G_perturbation)
+        self.epsilon_smooth = bool(epsilon_smooth)
+
+        # Replace neural heads with simple learnable parameters
+        # O parameters: only O[1] is learnable (O[0] is always frozen to 1.0)
+        self.O_params = nn.Parameter(
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:76]: the value of self.O_params should reflect the state of O used for the DMU calculations for better code readability and debugging. At the moment, it's quite confusing. Currently, if freeze_O=False, the value of O[0] = 0 which isn't true for the O used in the DMU equation (where O[0] = 1). Similarly, if freeze_O=True, then O_params=[0,0.5] which is not the O used in the calculations.

File: stable_nalu/layer/dmu.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        self._mag_max = 1e6
+        self._log_lim = math.log(self._mag_max) - 1.0
+
+    def _setup_freeze_G_gradients(self):
+        """Set up gradient masking for freeze_G mode."""
+        # We'll handle this in the forward pass by creating a new tensor
+        # that zeros out gradients for the first parameter
+        pass
+
+    def reset_parameters(self) -> None:
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:102]: To ensure gradients are never passed to the optimizer is O is frozen use: 
```
if freeze_O:
    self.O_params.requires_grad_(False)
```

File: stable_nalu/layer/dmu.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        # Initialize G parameters to neutral (0 logits = equal probability)
+        with torch.no_grad():
+            self.G_params.data = torch.tensor(
+                [-self.G_log_bias, self.G_log_bias], dtype=torch.float32
+            )
+
+    def predict_dmu_weights(self, device, dtype, B: int):
+        """Generate O and G weights using simple learnable parameters (input-independent)."""
+
+        # O weights: O[0] is always frozen to 1.0, O[1] depends on freeze_O setting
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:121]: Since O[0] is always 1 consider having a scalar o1 param to learn instead of O_params, i.e.,
self.o1 = nn.Parameter(torch.tensor(float(init_o)))

and then use stack to combine with the value of O[0] which is just a tensor of 1's like you have been doing.

File: stable_nalu/layer/dmu.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        self.in_features = in_features
+        self.out_features = out_features
+
+        # Fixed architecture: depth=1, 2 inputs, 1 output
+        self.num_initial_nodes = 2
+        self.G_log_bias = G_log_bias
+        self.G_temperature = G_temperature
+
+        self.op = op
+        self.freeze_G = bool(freeze_G)
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:68]: self.freeze_G does not freeze G_params, instead it modifies the value of O and creates the values for G_lin and G_log. Makes more sense to consider renaming freeze_G → learnable_O_only

File: stable_nalu/layer/dmu.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        """Calculate sparsity error based on G (gating) weights.
+
+        Args:
+            operation: The arithmetic operation (kept for API compatibility, not used)
+
+        Returns:
+            sparsity_error: min(|G|, |1-G|) measuring distance from G to discrete values {0,1}
+        """
+        # Get the current raw G weights directly from parameters
+        g_probs = torch.softmax(self.G_params, dim=0)
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:279]: also use `temperature` to be consistent with calculation used in forward

File: stable_nalu/layer/dmu.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+            if self.freeze_G:
+                # For freeze_G mode: O[1] starts at 0 and learns within [-1, 1]
+                self.O_params.data = torch.tensor([0.0, 0.0])  # [unused, O[1]]
+            else:
+                # Normal mode: O[1] starts at 0.5 (neutral initialization)
+                self.O_params.data = torch.tensor([0.0, 0.5])  # [unused, O[1]]
+
+        # Initialize G parameters to neutral (0 logits = equal probability)
+        with torch.no_grad():
+            self.G_params.data = torch.tensor(
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:114]: Something to compare/consider - Since there's only 2 options (either take log path or linear path), consider learning a scalar rather than a vector and use sigmoid to calculate the probability for taking the linear path, and then inferring the other path's value using `1 - prob(linear path)`. 

Though switching from softmax to sigmoid will affect gradient updates so will impact convergence in some way.

File: stable_nalu/layer/dmu.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+                elif self.op in ["mul", "div"]:
+                    # Log operations: G=0.0 for linear, G=1.0 for log
+                    G_lin = (
+                        torch.zeros(B, device=device, dtype=dtype) + self.G_perturbation
+                    )
+                    G_log = (
+                        torch.ones(B, device=device, dtype=dtype) - self.G_perturbation
+                    )
+
+        return O, G_lin, G_log
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:202]: return the expanded versions of G_lin and G_log e.g., G_lin.unsqueeze(-1)  and remove the G_lin_expanded, G_log_expanded variables in forward to reducecode duplication.

File: stable_nalu/layer/dmu.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+        m_log = torch.clamp(m_log, min=-self._log_lim, max=self._log_lim)
+        V_mag_new = torch.exp(m_log)
+
+        # Final result
+        final_value = (V_sign_new * V_mag_new).squeeze(-1)  # [B]
+
+        return final_value.unsqueeze(-1)  # [B, 1] for compatibility
+
+    def calculate_sparsity_error(self, operation: str) -> float:
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:269]: consider adding support for regularisation to enforce interpretable parameters like in other NALM modules

File: stable_nalu/layer/dmu.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+        # Get operation weights and domain mixing parameters
+        O, G_lin, G_log = self.predict_dmu_weights(device, dtype, B)
+
+        # Set up working arrays (simplified for depth=1)
+        working_mag = torch.abs(input)  # [B, 2]
+        working_sign = init_sign  # [B, 2]
+
+        # Compute linear and log domain results using original logic
+        R_lin = torch.sum(O * input, dim=-1, keepdim=True)  # [B, 1]
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:230]: typo in paper in sec 5.1 for calculation of y_lin and in the algorithm for the calculation of r_lin. Both are using `x_mag` instead of **x**.

File: stable_nalu/layer/dmu.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+        writer: str = None,
+        name: str | None = None,
+        op: str = None,
+        freeze_G: bool = False,
+        freeze_O: bool = False,
+        unfreeze_eval: bool = False,
+        G_perturbation: float = 0.0,
+        epsilon_smooth: bool = False,
+        G_log_bias: float = 0,
+        G_temperature: float = 0.1,
```
Comment (bmistry4) [inline stable_nalu/layer/dmu.py:43]: currently no support to pass most these arguments in when running experiments. Either add the support or move them to be initalised in the init. E.g., instead of self.G_log_bias = G_log_bias do self.G_log_bias = 0.

File: stable_nalu/layer/__init__.py [hunk 13]
```diff
@@ -1,5 +1,6 @@
 
 from .basic import BasicLayer, BasicCell
+from .dmu import DMULayer
```
Comment (bmistry4) [inline stable_nalu/layer/__init__.py:3]: add to the README.md file a bullet point for the dmu layer (after the MC-FC bullet pont).

**Review by bmistry4 — CHANGES_REQUESTED** [review]
Please see the requested changes and comments. 

Overall, good initial implementation and the layer functionally seems to work, but needs some cleaning up to make the code easier to follow. In particular, the parameters and code relevant to setting G and O :)
